added bsonBatchSize to config

cryogenian opened the pull request:
I'll just leave it here 🙂 

This adds optional field for maximum bsons size that can be put into queue. 

+ I'm not totally sure that this has any sense actually, the size of bsons is much smaller than size of equivalent json, and this basically is magical number that somehow connected with document size and RAM. 
+ This doesn't solve https://app.clubhouse.io/data/story/4265/failure-to-discover-structure-on-mongodb-connector, and I suppose that 2048 queue is OK (in particular for the case described in the issue, I have ~890Mb RAM used before OOM, for other collections everything works fine with 5Gb RAM used)

wemrysi commented on datasource/src/main/scala/quasar/physical/mongo/Mongo.scala:
```diff
@@ -69,46 +72,73 @@ class Mongo[F[_]: MonadResourceErr : ConcurrentEffect] private[Mongo](client: Mo
+    val max = if (maxMemory > MaxBsonBatch) MaxBsonBatch else maxMemory.toLong
+
+    getStats map { doc =>
+      val avgObjSize = doc.get("avgObjSize").fold(0)(_.asInt32.getValue)
```
Safer to default to `1` in order to avoid div by zero error.

wemrysi commented on datasource/src/main/scala/quasar/physical/mongo/Mongo.scala:
```diff
@@ -69,46 +72,73 @@ class Mongo[F[_]: MonadResourceErr : ConcurrentEffect] private[Mongo](client: Mo
+  val DefaultBsonBatch: Int = 1024 * 128
+  val MaxBsonBatch: Long = 128L * 1024L * 1024L
+  val DefaultQueueSize: Long = 64L
+  val MaxQueueSize: Long = 2048L
```
I don't think we need a max queue size given the batch size limits.

wemrysi commented on datasource/src/main/scala/quasar/physical/mongo/Mongo.scala:
```diff
@@ -69,46 +72,73 @@ class Mongo[F[_]: MonadResourceErr : ConcurrentEffect] private[Mongo](client: Mo
 }
 
 object Mongo {
+  val DefaultBsonBatch: Int = 1024 * 128
```
With absolutely zero evidence to support this suggestion, I think defaulting to 2MB would be a bit nicer for throughput as 128k seems pretty small.

wemrysi commented on datasource/src/main/scala/quasar/physical/mongo/MongoConfig.scala:
```diff
@@ -20,18 +20,18 @@ import slamdata.Predef._
 import argonaut._, Argonaut._
 
-final case class MongoConfig(connectionString: String)
+final case class MongoConfig(connectionString: String, bsonBatchSize: Option[Int])
```
How about naming this field something like `resultBatchSizeBytes` for a bit more self-documentation?

cryogenian commented:
@wemrysi 

+ `bsonBatchSize` -> `resultBatchSizeBytes`
+ removed redundant constant
+ `fold(0)` -> `fold(1)`
+ moved `val` out of method 

I didn't modify default batch size because 128K is the size of _bson_, json constructed from it is huge, i.e. tens of Mbs.

wemrysi commented on datasource/src/main/scala/quasar/physical/mongo/Mongo.scala:
```diff
@@ -119,12 +121,9 @@ class Mongo[F[_]: MonadResourceErr : ConcurrentEffect] private[Mongo](client: Mo
-      val avgObjSize = doc.get("avgObjSize").fold(0)(_.asInt32.getValue)
-      val res = max / avgObjSize.toLong
-      if (res < 1L) 1L else if (res < MaxQueueSize) res else MaxQueueSize
+      val avgObjSize = doc.get("avgObjSize").fold(1)(_.asInt32.getValue)
```
Instead of defaulting to a 1-byte average, what do you think about just having a default queue size of, say, 256 elements and using that if we don't know the `avgObjSize`.

I think it's good to be pessimistic here as we don't know anything about the data and in the current case we'll have a queue of `maximumBatchBytes` elements if we are unable to obtain the object size.

wemrysi commented:
@cryogenian Good call re the max bytes being in bson, sounds good :+1:

cryogenian commented:
@wemrysi 
+ Made default queue size 256
+ If there is no `avgObjSize` field or it's not int or long set queue size to 256
